File: JavaScript/380-Insert Delete GetRandom O(1).js

var RandomizedSet = function() {
    this.hmap = new Map();
    this.arr = new Array();
};

/** 
 * @param {number} val
 * @return {boolean}
 */
RandomizedSet.prototype.insert = function(val) {
    if (!this.hmap.has(val)){
        let n = this.arr.length;
        this.arr.push(val);
        this.hmap.set(val, n);
        return true;
    }else{
        return false;
    }
};

/** 
 * @param {number} val
 * @return {boolean}
 */
RandomizedSet.prototype.remove = function(val) {
    if (this.hmap.has(val)){
        let n = this.arr.length, currIdx = this.hmap.get(val);
        this.arr[currIdx] = this.arr[n-1];
        this.arr.pop();
        this.hmap.delete(val);
        this.hmap.set(this.arr[currIdx], currIdx)
        return true;
    }else{
        return false;
    }
};

/**
 * @return {number}
 */
RandomizedSet.prototype.getRandom = function() {
    return this.arr[Math.floor(Math.random()*this.arr.length)]
};

File: C++/380-Insert Delete GetRandom O(1).cpp
class RandomizedSet {
public:
    unordered_map<int, int> hmap;
    vector<int> res;
    RandomizedSet() {
        
    }
    
    bool insert(int val) {
        if (hmap.find(val) != hmap.end()) return false;
        res.push_back(val);
        hmap[val] = res.size() - 1;
        return true;
    }
    
    bool remove(int val) {
        if (hmap.find(val) == hmap.end()) return false;

        res[hmap[val]] = res.back();
        hmap[res.back()] = hmap[val];
        hmap[val] = res.size()-1;

        res.pop_back();
        hmap.erase(val);

        return true;
    }
    
    int getRandom() {
        return res[rand() % res.size()];
    }
};

/**
 * Your RandomizedSet object will be instantiated and called as such:
 * RandomizedSet* obj = new RandomizedSet();
 * bool param_1 = obj->insert(val);
 * bool param_2 = obj->remove(val);
 * int param_3 = obj->getRandom();
 */